lake/commits/store.go

```diff
@@ -0,0 +1,279 @@
+		paths:     make(map[ksuid.KSUID][]ksuid.KSUID),
+	}, nil
+}
+
+func (s *Store) Get(ctx context.Context, commit ksuid.KSUID) (*Object, error) {
+	if o, ok := s.cache[commit]; ok {
+		return o, nil
+	}
+	r, err := s.engine.Get(ctx, s.pathOf(commit))
+	if err != nil {
+		return nil, err
+	}
+	o, err := DecodeObject(r)
+	if err == ErrBadCommitObject {
+		err = fmt.Errorf("system error: %s: %w", s.pathOf(commit), ErrBadCommitObject)
+	}
+	if closeErr := r.Close(); err == nil {
+		err = closeErr
+	}
+	if err != nil {
+		return nil, err
+	}
+	//XXX do a length check on the table to limit memory use
+	s.cache[commit] = o
+	return o, nil
+}
+
+func (s *Store) pathOf(commit ksuid.KSUID) *storage.URI {
+	return s.path.AppendPath(commit.String() + ".zng")
+}
+
+func (s *Store) Put(ctx context.Context, o *Object) error {
+	b, err := o.Serialize()
+	if err != nil {
+		return err
+	}
+	return storage.Put(ctx, s.engine, s.pathOf(o.Commit), bytes.NewReader(b))
+}
+
+// DANGER ZONE - objects should only be removed when GC says they are not used.
+func (s *Store) Remove(ctx context.Context, o *Object) error {
+	return s.engine.Delete(ctx, s.pathOf(o.Commit))
+}
+
+func (s *Store) Snapshot(ctx context.Context, leaf ksuid.KSUID) (*Snapshot, error) {
+	if snap, ok := s.snapshots[leaf]; ok {
+		return snap, nil
+	}
+	var objects []*Object
+	var base *Snapshot
+	at := leaf
+	for at != ksuid.Nil {
+		if snap, ok := s.snapshots[at]; ok {
+			base = snap
+			break
+		}
+		o, err := s.Get(ctx, at)
+		if err != nil {
+			return nil, err
+		}
+		objects = append(objects, o)
+		at = o.Parent
+	}
+	var snap *Snapshot
+	if base == nil {
+		snap = NewSnapshot()
+	} else {
+		snap = base.Copy()
+	}
+	for k := len(objects) - 1; k >= 0; k-- {
+		for _, action := range objects[k].Actions {
+			if err := PlayAction(snap, action); err != nil {
+				return nil, err
+			}
+		}
+	}
+	s.snapshots[leaf] = snap
+	return snap, nil
+}
+
+// Path return the entire path from the commit object to the root
+// in leaf to root order.
+func (s *Store) Path(ctx context.Context, leaf ksuid.KSUID) ([]ksuid.KSUID, error) {
+	if leaf == ksuid.Nil {
+		return nil, errors.New("no path for nil commit ID")
+	}
+	if path, ok := s.paths[leaf]; ok {
+		return path, nil
+	}
+	path, err := s.PathRange(ctx, leaf, ksuid.Nil)
+	if err != nil {
+		return nil, err
+	}
+	s.paths[leaf] = path
+	return path, nil
+}
+
+func (s *Store) PathRange(ctx context.Context, from, to ksuid.KSUID) ([]ksuid.KSUID, error) {
+	var path []ksuid.KSUID
+	at := from
+	for at != ksuid.Nil {
+		if cache, ok := s.paths[at]; ok {
+			for _, id := range cache {
+				path = append(path, id)
+				if id == to {
+					break
+				}
+			}
+			break
+		}
+		path = append(path, at)
+		o, err := s.Get(ctx, at)
+		if err != nil {
+			return nil, err
+		}
+		if at == to {
+			break
+		}
+		at = o.Parent
+	}
+	return path, nil
+}
+
+func (s *Store) GetBytes(ctx context.Context, commit ksuid.KSUID) ([]byte, *Commit, error) {
```

Comment:
I would move this next to Get and Get should probably just call GetByes and return the Commit. Also should this populate the cache?

Reply:
The cache doesn't save the zng buffer.  These are two different models.  I think it will change anyway a bit when we persist the snapshots so let's just leave this for now.